va/http.go

```diff
@@ -0,0 +1,424 @@
+	ctx context.Context,
+	host string,
+	path string) ([]byte, []core.ValidationRecord, error) {
+	deadline, ok := ctx.Deadline()
+	if !ok {
+		// Shouldn't happen: All requests should have a deadline by this point.
+		deadline = time.Now().Add(100 * time.Second)
+	} else {
+		// Set the context deadline slightly shorter than the HTTP deadline, so we
+		// get a useful error rather than a generic "deadline exceeded" error. This
+		// lets us give a more specific error to the subscriber.
+		deadline = deadline.Add(-10 * time.Millisecond)
+	}
+	ctx, cancel := context.WithDeadline(ctx, deadline)
+	defer cancel()
+
+	// Create a target for the host, port and path
+	target, err := va.newHTTPValidationTarget(ctx, host, va.httpPort, path)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Set up the initial validation request and a base validation record
+	initialReq, baseRecord, err := va.setupHTTPValidation(ctx, nil, target)
+	if err != nil {
+		return nil, []core.ValidationRecord{}, err
+	}
+
+	va.log.AuditInfof("Attempting to validate HTTP-01 for %q with GET to %q",
+		initialReq.Host, initialReq.URL.String())
+
+	// Create a closure around records & numRedirects we can use with a HTTP
+	// client to process redirects per our own policy (e.g. resolving IP
+	// addresses explicitly, not following redirects to ports != [80,443], etc)
+	records := []core.ValidationRecord{baseRecord}
+	numRedirects := 0
+	processRedirect := func(req *http.Request, via []*http.Request) error {
+		va.log.Infof("processing a HTTP redirect from the server to %q\n", req.URL.String())
+		// Only process up to maxRedirect redirects
+		if numRedirects > maxRedirect {
+			return berrors.ConnectionFailureError("Too many redirects")
+		}
+		numRedirects++
+		va.metrics.http01Redirects.Inc()
+
+		// Extract the redirect target's host and port. This will return an error if
+		// the redirect request scheme, host or port is not acceptable.
+		redirHost, redirPort, err := va.extractRequestTarget(req)
+		if err != nil {
+			return err
+		}
+
+		redirPath := req.URL.Path
+		// Setup a validation target for the redirect host. This will resolve IP
+		// addresses for the host explicitly.
+		redirTarget, err := va.newHTTPValidationTarget(ctx, redirHost, redirPort, redirPath)
+		if err != nil {
+			return err
+		}
+
+		// Mutate the existing redirect request to use a URL and Host based on the
+		// explicitly resolved target IPs. This will also give us a validationRecord
+		// for the redirect which we should append to the records.
+		_, redirRecord, err := va.setupHTTPValidation(ctx, req, redirTarget)
+		records = append(records, redirRecord)
+		if err != nil {
+			return err
+		}
+		va.log.Infof("following redirect to host %q url %q\n", req.Host, req.URL.String())
+		return nil
+	}
+
+	// Create a new HTTP client and check HTTP redirects it encounters with
+	// processRedirect
+	client := newHTTPClient()
+	client.CheckRedirect = processRedirect
```

Comment:
Since we'll always assign a `CheckRedirect` to the client that comes out of `newHTTPClient()`, want to make it an argument and just assign it inside `newHTTPClient()`?

Reply:
That's a good idea :+1: I wanted to keep the closure in this function and didn't think to pass it as an argument to `newHTTPClient`. Thanks for the suggestion!